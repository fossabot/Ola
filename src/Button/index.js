import React from 'react'
import { default as PT } from 'prop-types'
import cx from 'classnames'
import { getElementType } from '../utils'

import Spinner from '../Spinner'
import Icon from '../Icon'

const Button = ({ variant, children, disabled, busy, className, iconLink, as, ...props }) => {
  const ElementType = getElementType(Button, { as: as, ...props })
  delete props['as']
  const styles = cx(
    'ola_button',
    { [`is-${variant}`]: variant },
    { 'is-busy': busy },
    { 'is-disabled': disabled && !busy },
    className
  )
  const isIcon = (variant === 'link' && iconLink)
  return (
    <ElementType className={styles} disabled={busy ? true : disabled} {...props}>
      {busy && <Spinner />}
      <span className="ola_button-text">{!busy ? children : busy}</span>
      {isIcon && <Icon name="iconLink" />}
    </ElementType>
  )
}

Button.defaultProps = {
  as: 'button',
  variant: null,
  busy: null,
  className: null,
  disabled: false,
  iconLink: false
}

Button.propTypes = {
  /** Render Button with any html tag */
  as: PT.string,
  /** Button variants */
  variant: PT.oneOf(['primary', 'secondary', 'destructive-primary', 'destructive', 'pro', 'link']),
  /** Text for loading state */
  busy: PT.oneOfType([
    PT.string,
    PT.bool
  ]),
  /** Extra className */
  className: PT.string,
  /** Childen nodes */
  children: PT.oneOfType([
    PT.string,
    PT.arrayOf(PT.node),
    PT.node
  ]).isRequired,
  /** Button is disabled */
  disabled: PT.bool,
  /**Icon is a Bolean  */
  iconLink: PT.bool
}

export default Button
