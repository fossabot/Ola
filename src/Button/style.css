.ola_button {
    --color: var(--accent);
    --color-hover: var(--accent-dark);
    --color-selected: var(--black);
    --background: transparent;
    --background-hover: var(--gray-xlight);
    --background-active: var(--accent-focus);
    --background-selected: var(--gray-xlight);
    --shadow-focus: var(--accent-focus);
    --border: var(--gray-light);
    --border-hover: currentColor;
    --border-selected: var(--gray-light);

    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    height: var(--size-8);
    min-width: var(--size-8);
    max-width: 100%;
    border: none;
    border-color: var(--border);
    background: none;
    color: var(--color);
    background-color: var(--background);
    cursor: pointer;
    border-radius: var(--radius-big);
    font: var(--font-callout);
    padding: 0 var(--column-gap-2);
    font-weight: var(--bold);
    transition-property: background-color, color, box-shadow;
    transition-duration: var(--duration);
    outline: 0;
    white-space: nowrap;
    text-decoration: none;
    box-sizing: border-box;
    vertical-align: middle;
    flex: 0 0 auto;

    &::-moz-focus-inner {
        border: 0;
    }

    &:hover {
        color: var(--color-hover);
        background-color: var(--background-hover);
        border-color: var(--border-hover);
    }

    &:focus {
        box-shadow: 0 0 0 4px var(--shadow-focus);
    }

    &:active {
        background-color: var(--background-active);
    }

    &.is-primary {
        --color: var(--white);
        --background: var(--accent);
        --color-hover: var(--white);
        --background-hover: var(--accent-dark);
        --shadow-focus: var(--accent-focus);
        --background-active: var(--black);
    }

    &.is-secondary {
        border-style: solid;
        border-width: 1px;
    }

    &.is-destructive-primary {
        --color: var(--white);
        --background: var(--error);
        --color-hover: var(--white);
        --background-hover: var(--error-dark);
        --shadow-focus: var(--error-focus);
        --background-active: var(--black);
    }

    &.is-destructive {
        --color: var(--error);
        --shadow-focus: var(--error-focus);
        --color-hover: var(--error-dark);
        --background-active: var(--error-focus);

        border-style: solid;
        border-width: 1px;
    }

    &.is-pro {
        --color: var(--white);
        --background: var(--pro);
        --color-hover: var(--white);
        --background-hover: var(--pro-dark);
        --shadow-focus: var(--pro-focus);
        --background-active: var(--black);
    }

    &.is-busy {
        color: var(--white);
        background-color: var(--gray);
        min-width: 0;
        cursor: wait;
        pointer-events: none;
        border: none;

        & .ola_spinner {
            position: relative;

            &:first-child {
                right: var(--size-6);
            }

            &:last-child {
                left: var(--size-6);
            }
        }
    }

    &.is-link {
        --color: var(--brand);
        --color-hover: var(--black);
        --background-hover: transparent;
        --shadow-focus: var(--brand-focus);
        --background-active: var(--brand-focus);

        cursor: pointer;
        min-width: 0;
        padding-left: 0;
        padding-right: 0;
        border-radius: 0;
        text-decoration: underline;
        text-underline-offset: 4px;
        text-decoration-thickness: 1px;

        &.is-disabled,
        &:disabled {
            color: var(--gray-light);
            text-decoration: none;
            background: none;
        }

        &.is-busy {
            color: var(--gray);
            text-decoration: none;
            background: none;
            cursor: wait;

            & .ola_spinner {
                position: static;

                &:first-child {
                    margin-right: var(--size-4);
                }

                &:last-child {
                    margin-left: var(--size-4);
                }
            }
        }
    }

    &.is-disabled,
    &:disabled {
        color: var(--gray-light);
        --background: var(--gray-xlight);
        --background-hover: var(--gray-xlight);
        --background-active: var(--gray-xlight);
        border: none;
        cursor: not-allowed;
        pointer-events: none;
        box-shadow: none;
    }

    &[aria-pressed="true"] {
        color: var(--color-selected);
        background-color: var(--background-selected);
        border-color: var(--border-selected);
        cursor: default;
        pointer-events: none;
    }
}

.ola_button-text {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    padding-top: 4px;
    padding-bottom: 4px;
}

.ola_button-icon {
    right: var(--size-4);
    position: relative;
}

summary:focus .ola_button {
    box-shadow: 0 0 0 4px var(--focus);
}

.ola_icon svg {
    vertical-align: middle;
}
