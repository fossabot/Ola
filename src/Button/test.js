import React from 'react'
import { Button } from './'
import renderer from 'react-test-renderer'

it('No variant button', () => {
  const tree = renderer
    .create(<Button>Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Primary button', () => {
  const tree = renderer
    .create(<Button variant="primary">Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Secondary button', () => {
  const tree = renderer
    .create(<Button variant="secondary">Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('PRO button', () => {
  const tree = renderer
    .create(<Button variant="pro">Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Primary destructive button', () => {
  const tree = renderer
    .create(<Button variant="destructive-primary">Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Destructive button', () => {
  const tree = renderer
    .create(<Button variant="destructive">Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Busy button', () => {
  const tree = renderer
    .create(<Button variant="primary" busy>Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Primary button with extra class', () => {
  const tree = renderer
    .create(<Button variant="primary" extraClass="is-extra">Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Disabled primary button', () => {
  const tree = renderer
    .create(<Button variant="primary" disabled>Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Disabled link button', () => {
  const tree = renderer
    .create(<Button variant="link" disabled>Click here</Button>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})