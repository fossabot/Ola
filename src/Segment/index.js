import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'

const Segment = ({ className, children, ...props }) => {
  const styles = cx('ola_segment', className)

  return (
    <nav className={styles} {...props}>
      { children }
    </nav>
  )
}

Segment.defaultProps = {
  className: null
}

Segment.propTypes = {
  /** Extra className */
  className: PT.string,
  /** Childen nodes */
  children: PT.oneOfType([
    PT.string,
    PT.arrayOf(PT.node),
    PT.node
  ]).isRequired
}

export default Segment
