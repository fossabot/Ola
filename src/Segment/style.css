.ola_segment {
    display: inline-flex;
    max-width: 100%;

    & > .ola_button {
        border-radius: 0;
        flex-shrink: 1;

        & + .ola_button {
            border-left: none;
        }
    }

    & > .ola_button:first-child {
        border-top-left-radius: var(--radius-big);
        border-bottom-left-radius: var(--radius-big);
    }
    & > .ola_button:last-child {
        border-top-right-radius: var(--radius-big);
        border-bottom-right-radius: var(--radius-big);
    }
}
