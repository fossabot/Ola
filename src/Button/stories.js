import React from 'react'
import Button from './'

export default {
  title: 'Button',
  component: Button,
  args: {
    children: 'Button',
  },
  argTypes: {
    onClick: { action: 'clicked' },
    onFocus: { action: 'focused' },
    busy: { control: 'text' },
    as: {
      control: {
        type: 'inline-radio',
        options: ['button', 'a', 'span'],
      },
    },
  },
}

export const Base = (args) => <Button {...args}>{args.children}</Button>
export const WithOlaIcon = (args) => (
  <Button variant="secondary" icon="add" {...args}>
    {args.children}
  </Button>
)
export const WithIcon = (args) => (
  <Button
    variant="secondary"
    icon={<img src="https://placeholder.pics/svg/20" />}
    {...args}
  >
    {args.children}
  </Button>
)
