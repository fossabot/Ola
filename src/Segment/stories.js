import React from 'react'
import Button from '../Button'
import Segment from './'

export default {
  title: 'Segment',
  component: Segment
}

export const Base = (args) => <Segment {...args}>
  <Button selected variant='secondary'>All keywords</Button>
  <Button selected={false} variant='secondary'>Only focus keywords</Button>
</Segment>

Base.argTypes = {
  children: { control: { disable: true }}
}