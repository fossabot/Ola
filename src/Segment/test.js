import React from 'react'
import Button from '../Button'
import Segment from './'
import renderer from 'react-test-renderer'

it('Segment with two buttons', () => {
  const tree = renderer
    .create(
      <Segment>
        <Button variant="secondary">Main button</Button>
        <Button variant="secondary">Secondary button</Button>
      </Segment>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it('Segment with two buttons and an extra class and one button pressed', () => {
  const tree = renderer
    .create(
      <Segment className="extra-class">
        <Button variant="secondary" selected>Main button</Button>
        <Button variant="secondary">Secondary button</Button>
      </Segment>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
