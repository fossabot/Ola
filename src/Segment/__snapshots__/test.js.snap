// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segment with two buttons 1`] = `
<nav
  className="ola_segment"
>
  <button
    aria-pressed={null}
    className="ola_button is-secondary"
    disabled={false}
  >
    <span
      className="ola_button-text"
    >
      Main button
    </span>
  </button>
  <button
    aria-pressed={null}
    className="ola_button is-secondary"
    disabled={false}
  >
    <span
      className="ola_button-text"
    >
      Secondary button
    </span>
  </button>
</nav>
`;

exports[`Segment with two buttons and an extra class and one button pressed 1`] = `
<nav
  className="ola_segment extra-class"
>
  <button
    aria-pressed="true"
    className="ola_button is-secondary"
    disabled={false}
  >
    <span
      className="ola_button-text"
    >
      Main button
    </span>
  </button>
  <button
    aria-pressed={null}
    className="ola_button is-secondary"
    disabled={false}
  >
    <span
      className="ola_button-text"
    >
      Secondary button
    </span>
  </button>
</nav>
`;
